server/mailet/rate-limiter/src/main/scala/org/apache/james/transport/mailets/PerSenderRateLimit.scala

```diff
@@ -0,0 +1,211 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ * http://www.apache.org/licenses/LICENSE-2.0                   *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.transport.mailets
+
+import java.time.Duration
+import java.time.temporal.ChronoUnit
+
+import com.google.common.annotations.VisibleForTesting
+import eu.timepit.refined.auto._
+import javax.inject.Inject
+import org.apache.james.core.MailAddress
+import org.apache.james.rate.limiter.api.Quantity.Quantity
+import org.apache.james.rate.limiter.api.{AcceptableRate, Quantity, RateExceeded, RateLimiter, RateLimiterFactoryProvider, RateLimitingKey, RateLimitingResult, Rule, Rules}
+import org.apache.james.util.DurationParser
+import org.apache.mailet.Mail
+import org.apache.mailet.base.GenericMailet
+import org.reactivestreams.Publisher
+import reactor.core.scala.publisher.{SFlux, SMono}
+
+case class KeyPrefix(value: String)
+
+sealed trait EntityType {
+  def asString(): String
+
+  def extractQuantity(mail: Mail): Quantity
+}
+
+case object Count extends EntityType {
+  override def asString(): String = "count"
+
+  override def extractQuantity(mail: Mail): Quantity = 1
+}
+case object RecipientsType extends EntityType {
+  override def asString(): String = "recipients"
+
+  override def extractQuantity(mail: Mail): Quantity = Quantity.liftOrThrow(mail.getRecipients.size())
+}
+case object Size extends EntityType {
+  override def asString(): String = "size"
+
+  override def extractQuantity(mail: Mail): Quantity = Quantity.liftOrThrow(mail.getMessageSize.toInt)
+}
+case object TotalSize extends EntityType {
+  override def asString(): String = "totalSize"
+
+  override def extractQuantity(mail: Mail): Quantity = Quantity.liftOrThrow((mail.getMessageSize * mail.getRecipients.size()).toInt)
```

Comment:
`mail.getMessageSize` returning  `long`. 
So `.toLong` is better?

Anw,
Max value of `Int` = 2147483647
2147483647 bytes = ~ 2147MB = ~ 2GB

Reply:
Fair. We may need to split `Quantity` in two: `Increments` and `AllowedQuantity`. 

Note that the underlying library `RateLimitJ` would anyway enforce int usage for the increments.... Not much we could do for this. (But the above proposition lifts the contstraint for the allowed quantity and is thus desirable).

Maybe this could be done in a third party PR...

Reply:
Or we can convert bytes number to KB/MB number before input to API.